src/group_theory/presented_group.lean

```diff
@@ -0,0 +1,73 @@
+/-
+Copyright (c) 2019 Michael Howes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Howes
+
+Defining a group given by generators and relations
+-/
+
+import group_theory.free_group group_theory.quotient_group
+
+variables {α : Type}
+
+/-- Given a set of relations, rels, over a type α, presented_group constructs the group with
+generators α and relations rels as a quotient of free_group α.-/
+def presented_group (rels : set (free_group α)) : Type :=
+    quotient_group.quotient $ group.normal_closure rels
+
+namespace presented_group
+
+instance (rels : set (free_group α)) : group (presented_group (rels)) :=
+quotient_group.group _
+
+/-- `of x` is the canonical map from α to a presented group with generators α. The term x is
+mapped to the equivalence class of the image of x in free_group α. -/
+def of {rels : set (free_group α)} (x : α) : presented_group rels :=
+quotient_group.mk (free_group.of x)
+
+namespace to_group
+
+/-
+Presented groups satisfy a universal property. If β is a group and f : α → β is a map such that the
+images of f satisfy all the given relations, then f extends uniquely to a group homomorphism from
+presented_group rels to β
+-/
+
+variables {β : Type} [group β] {f : α → β} {rels : set (free_group α)}
+
+local notation `F` := free_group.to_group f
+
+variable (h : ∀ r ∈ rels, F r = 1)
+
+lemma closure_subset_kernel : group.normal_closure rels ⊆ is_group_hom.ker F :=
+group.normal_closure_subset (λ x w, (is_group_hom.mem_ker F).2 (h x w))
+
+lemma closure_mapsto_zero : ∀ x ∈ group.normal_closure rels, F x = 1 :=
+λ x w, (is_group_hom.mem_ker F).1  ((closure_subset_kernel h) w)
+
+/-- The extension of a map f : α → β that satisfies the given relations to a group homomorphism
+from presented_group rels → β. -/
+def to_group : presented_group rels → β :=
+quotient_group.lift (group.normal_closure rels) F (closure_mapsto_zero h)
+
+instance to_group.is_group_hom : is_group_hom (to_group h) :=
+quotient_group.is_group_hom_quotient_lift _ _ _
+
+@[simp] lemma to_group.of {x : α} : to_group h (of x) = f x := free_group.to_group.of
```

Comment:
All of these lemmas are in the `to_group` namespace twice. The full name of this will be `presented_group.to_group.to_group.of`. Also, `to_group` has the full name `presented_group.to_group.to_group`

Reply:
Thanks for pointing this out. I have changed the `to_group` namespace to a `to_group` section. I was trying to copy `group_theory/free_group` but didn't notice that they used a `to_group` section rather than a `to_group` namespace.